### ISSUE TITLE
[Quest][Northrend] The Grand Melee

### ISSUE THREAD

**Issue by humapal** [dalaranwow/dalaran-wow#4692]
[//]: # (INFORMATION ABOUT TAGS)
[//]: # (The black and yellow tag is unique. There can only be one label assigned to them - one black and one yellow)
[//]: # (Priority - ... tags are assigned by team members)
[//]: # (Your issue will be reviewed by testers and the developers will fix it in the future.)
[//]: # (Only once the issue got the label [Fixed - On Live Server] you can find the fix on your Gameserver.)
[//]: # (You can increase the relevance of your issue by adding a thumb-up emote http://i.imgur.com/RFouVVi.png)

[//]: # (***************************************************************)
[//]: # (** !!!!!!!!!!!!!DON'T DELETE THIS TEMPLATE ELSE YOUR ISSUE WILL BE CLOSED!!!!!!!!!!!! **)
[//]: # (***************************************************************)

**Current behaviour**: Tell us what happens in detail (Where, When, What, Why, Who)
You challenge each valiant to a fight but it does not give you credit when you defeat them.

**Expected behaviour**: Tell us what should happen instead
You should receive a credit for each time you beat the valiant in training.

You can see from the picture that I bested each valiant of every Alliance faction
but in the quest tracker I received no credit for any of them.  

![grandmelee](https://cloud.githubusercontent.com/assets/26784243/25570536/c5f1c026-2deb-11e7-863c-85421dc1b7e8.jpg)

**Steps to reproduce the problem**:

1.  Ride an Elekk  
2. Challenge any Alliance Valiant
3. Beat them and see if it gives you credit.

**Include proofs for this behaviour**

You can research this on several pages like http://wowhead.com or http://wowwiki.wikia.com or http://www.youtube.com/
Issues without proofs will be closed and tag On Hold will be added until the proof is provided.

**Include the ID for the game objects, npcs (creatures,pets,minions), spells, items, quests, instances, zones, achievements, skills**

You can research this on several pages like http://wotlk.openwow.com or http://wowhead.com/

**Include Screenshots from the issue if necessary**

**Comment by TeridanDW** [comment]
Hello @humapal , is this issue happening on Algalon? 
Did **you** do the "killing" blow of the creature while riding a mount and having a lance equiped as your weapon?

**Comment by humapal** [comment]
1. You can't challenge a valiant unless you are mounted.
2. You can't mount unless you have a lance equipped.
3. You can see that I am mounted on the Elekk in the picture and that I had 5 Killing blows
in 4 minutes and 57 seconds and you would never be able to do that if you weren't mounted without a lance.  That in itself should an achievement!!

So, the answer to your question is yes.  I am an argent champion with 4 out of 5 of the Alliance factions
and I am working on the last faction of Exodar.    I have always had problems with not getting credits for each kill but you if do it long enough you will eventually  get all of your credits.   The Eleek has been the only one not to grant a credit even though I have killed over 40 valiants.

**Comment by humapal** [comment]
![grand melee 2](https://cloud.githubusercontent.com/assets/26784243/25600981/16d1a3fe-2eae-11e7-9559-fcf6235c6601.jpg)

After 16 kills using a Stormwind Mount and Lance received no credit for any of the kills.  Although you can see that I bested 4 of them.

**Comment by TeridanDW** [comment]
Hello, I've been "digging" around this issue all day and now I can confirme that there is a problem with quest credits (not been granted in all cases) when a player beats a valiant/champion. 

Few clarifications before I go to the point:
 

1. - What kind of mount the player rides (house,tireg, etc...) does _**not**_ matter (it is not relevent to the script)
 

2. - When a valiant/champion is defeated the server will cast 2 spells on the player, 1-st is the quest credit (player creates an item in their inventory (free space in the inventory is a _**MUST**_ to get it)) and the 2-nd is the debuff that player gets after defeating the said creature (based on race).

Regarding the 1-st spell:
 - http://www.wowhead.com/spell=63060/quest-credit-trigger-01 and is been used by all valiants (all races A&H)  
 - http://www.wowhead.com/spell=63597/quest-credit-trigger-01 and is been used by all champions (all races A&H) 

Regarding the 2-nd spell - it's been used by valiants and champions based on race (so Human valiants and champions use the same spell, Orc valiants and champions are use the same spell, etc...) and the spells are:
 - Stormwind - http://www.wowhead.com/spell=64836/dnd-bested-trigger-stormwind
 - Darnassus - http://www.wowhead.com/spell=64831/dnd-bested-trigger-darnassus
 - Exodar - http://www.wowhead.com/spell=64838/dnd-bested-trigger-exodar
 - Ironforge - http://www.wowhead.com/spell=64832/dnd-bested-trigger-ironforge
 - Gromeregan - http://www.wowhead.com/spell=64830/dnd-bested-trigger-gnomeregan
 - Silvermoon - http://www.wowhead.com/spell=64835/dnd-bested-trigger-silvermoon
 - Thunder Bluff - http://www.wowhead.com/spell=64837/dnd-bested-trigger-thunder-bluff
 - Orgrimmar - http://www.wowhead.com/spell=64833/dnd-bested-trigger-orgrimmar
 - Sen'jin - http://www.wowhead.com/spell=64834/dnd-bested-trigger-senjin
 - Undercity - http://www.wowhead.com/spell=64839/dnd-bested-trigger-undercity

3. - Here are the entries of the valiants and champions:

 **Valiants:**
 - http://www.wowhead.com/npc=33383/thunder-bluff-valiant
 - http://www.wowhead.com/npc=33384/undercity-valiant
 - http://www.wowhead.com/npc=33306/orgrimmar-valiant
 - http://www.wowhead.com/npc=33285/senjin-valiant
 - http://www.wowhead.com/npc=33382/silvermoon-valiant
 - http://www.wowhead.com/npc=33561/stormwind-valiant
 - http://www.wowhead.com/npc=33564/ironforge-valiant
 - http://www.wowhead.com/npc=33558/gnomeregan-valiant
 - http://www.wowhead.com/npc=33559/darnassus-valiant
 - http://www.wowhead.com/npc=33562/exodar-valiant

**Champions:**
 - http://www.wowhead.com/npc=33740/gnomeregan-champion
 - http://www.wowhead.com/npc=33743/ironforge-champion
 - http://www.wowhead.com/npc=33747/stormwind-champion
 - http://www.wowhead.com/npc=33738/darnassus-champion
 - http://www.wowhead.com/npc=33739/exodar-champion
 - http://www.wowhead.com/npc=33749/undercity-champion
 - http://www.wowhead.com/npc=33745/senjin-champion
 - http://www.wowhead.com/npc=33744/orgrimmar-champion
 - http://www.wowhead.com/npc=33748/thunder-bluff-champion
 - http://www.wowhead.com/npc=33746/silvermoon-champion


To the point now, player will almost always fails to get credit if he:
 - engage a valiant while riding a mount 
 - drop the valiant to ~ 2000 hp
 - make sure the valiant doesn't have a charge of his shield
 - "finish" the valiant with spell http://www.wowhead.com/spell=62960/charge
 - notice you don't get spell 1 and 2 been casted on you

@humapal I couldn't reproduce your issue where you don't get credit but gets the debuff. I always either get both or none of them. Any tips how to do it?

After doing all of this research i did try to fix the issue myself and made a SAI for the Stormwind valiants (only for them) and with my version I always got the 2 spells casted on me, however I think that happend because of changes made by TC. Possible places where the problem(s) might be (my guess):
 - spell http://www.wowhead.com/spell=62960/charge - I would like to point out that this spell breaks the mini event i made (when I killed the valiant with that spell), which causes the valiant to instantly evades, after he kills the player, instead of doing the minievet. This does _**NOT**_ happen if I kill the valiant with spell http://www.wowhead.com/spell=62544/thrust or http://www.wowhead.com/spell=62575/shield-breaker
 - spell http://www.wowhead.com/spell=63060/quest-credit-trigger-01
 - spell http://www.wowhead.com/spell=64836/dnd-bested-trigger-stormwind

Here is the overall improved version of the Stormwind's valiants (has mini event on valiant win/lose/evade, fixed some visual issues, removed gossip flag on script start and added it back on creature reset):
``` sql
-- Stormwind Valiant SAI
SET @ENTRY := 33561;
UPDATE `creature_template` SET `AIName`="SmartAI" WHERE `entry`=@ENTRY;
DELETE FROM `smart_scripts` WHERE `entryorguid`=@ENTRY AND `source_type`=0;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(@ENTRY,0,0,23,25,0,100,512,0,0,0,0,28,64223,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Reset - Remove Aura '[DND] Defend Aura (10 seconds)'"),
(@ENTRY,0,1,2,62,0,100,512,10471,0,0,0,72,0,0,0,0,0,0,7,0,0,0,0,0,0,0,"Stormwind Valiant - On Gossip Option 0 Selected - Close Gossip"),
(@ENTRY,0,2,3,61,0,100,512,10471,0,0,0,64,1,0,0,0,0,0,7,0,0,0,0,0,0,0,"Stormwind Valiant - On Gossip Option 0 Selected - Store Targetlist"),
(@ENTRY,0,3,0,61,0,100,512,10471,0,0,0,80,@ENTRY*100+00,2,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Gossip Option 0 Selected - Run Script"),
(@ENTRY,0,4,0,61,0,100,512,0,33561,0,0,11,63010,0,0,0,0,0,22,0,0,0,0,0,0,0,"Stormwind Valiant - On Text 0 Over - Cast 'Charge'"),
(@ENTRY,0,5,0,0,0,100,512,0,0,2000,2000,11,62544,0,0,0,0,0,2,0,0,0,0,0,0,0,"Stormwind Valiant - In Combat - Cast 'Thrust'"),
(@ENTRY,0,6,0,9,0,100,512,5,5,9000,10000,11,62575,0,0,0,0,0,2,0,0,0,0,0,0,0,"Stormwind Valiant - Within 5-10 Range - Cast 'Shield-Breaker'"),
(@ENTRY,0,7,0,0,0,100,512,9000,11000,9000,10000,114,0,0,0,0,0,0,1,0,0,0,0,12,0,0,"Stormwind Valiant - In Combat"),
(@ENTRY,0,8,9,9,0,100,512,9,15,6000,6000,11,63010,0,0,0,0,0,2,0,0,0,0,0,0,0,"Stormwind Valiant - Within 9-15 Range - Cast 'Charge'"),
(@ENTRY,0,9,0,61,0,100,512,9,15,6000,6000,11,62575,0,0,0,0,0,2,0,0,0,0,0,0,0,"Stormwind Valiant - Within 9-15 Range - Cast 'Shield-Breaker'"),
(@ENTRY,0,10,0,2,0,100,513,0,1,60000,60000,80,@ENTRY*100+01,2,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - Between 0-1% Health - Run Script (No Repeat)"),
(@ENTRY,0,15,16,5,0,100,512,60000,60000,1,0,1,1,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Killed Unit - Say Line 1"),
(@ENTRY,0,16,17,61,0,100,512,60000,60000,1,0,27,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Killed Unit - Combat Stop"),
(@ENTRY,0,17,19,61,0,100,512,60000,60000,1,0,41,5000,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Killed Unit - Despawn In 5000 ms"),
(@ENTRY,0,18,4,52,0,100,512,0,33561,0,0,49,0,0,0,0,0,0,22,0,0,0,0,0,0,0,"Stormwind Valiant - On Text 0 Over - Start Attacking"),
(@ENTRY,0,19,0,61,0,100,512,60000,60000,1,0,102,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Killed Unit - Set Health Regeneration Off"),
(@ENTRY,0,20,21,21,0,100,512,0,0,0,0,2,35,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Reached Home - Set Faction 35"),
(@ENTRY,0,21,22,61,0,100,512,0,0,0,0,78,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Reached Home - Reset All Scripts"),
(@ENTRY,0,22,0,61,0,100,512,0,0,0,0,82,1,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Reached Home - Add Npc Flag Gossip"),
(@ENTRY,0,23,0,61,0,100,512,0,0,0,0,102,1,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Reset - Set Health Regeneration On");


-- Actionlist SAI
SET @ENTRY := 3356100;
DELETE FROM `smart_scripts` WHERE `entryorguid`=@ENTRY AND `source_type`=9;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(@ENTRY,9,1,0,0,0,100,0,0,0,0,0,83,1,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Remove Npc Flag Gossip"),
(@ENTRY,9,2,0,0,0,100,0,0,0,0,0,42,0,1,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Set Invincibility Hp 1%"),
(@ENTRY,9,3,0,0,0,100,0,0,0,0,0,59,1,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Set Run On"),
(@ENTRY,9,4,0,0,0,100,0,0,0,0,0,11,62719,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Cast 'Defend'"),
(@ENTRY,9,5,0,0,0,100,0,0,0,0,0,11,62719,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Cast 'Defend'"),
(@ENTRY,9,6,0,0,0,100,0,0,0,0,0,11,62719,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Cast 'Defend'"),
(@ENTRY,9,7,0,0,0,100,0,0,0,0,0,114,0,0,0,0,0,0,1,0,0,0,0,20,0,0,"Stormwind Valiant - On Script"),
(@ENTRY,9,8,0,0,0,100,0,0,0,0,0,59,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Set Run Off"),
(@ENTRY,9,9,0,0,0,100,0,0,0,0,0,11,64223,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Cast '[DND] Defend Aura (10 seconds)'"),
(@ENTRY,9,10,0,0,0,100,0,0,0,0,0,1,0,4000,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Say Line 0"),
(@ENTRY,9,11,0,0,0,100,0,3500,3500,0,0,2,14,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Set Faction 14");

-- Actionlist SAI
SET @ENTRY := 3356101;
DELETE FROM `smart_scripts` WHERE `entryorguid`=@ENTRY AND `source_type`=9;
INSERT INTO `smart_scripts` (`entryorguid`,`source_type`,`id`,`link`,`event_type`,`event_phase_mask`,`event_chance`,`event_flags`,`event_param1`,`event_param2`,`event_param3`,`event_param4`,`action_type`,`action_param1`,`action_param2`,`action_param3`,`action_param4`,`action_param5`,`action_param6`,`target_type`,`target_param1`,`target_param2`,`target_param3`,`target_x`,`target_y`,`target_z`,`target_o`,`comment`) VALUES
(@ENTRY,9,0,0,0,0,100,0,0,1,60000,60000,2,35,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Set Faction 35"),
(@ENTRY,9,1,0,0,0,100,0,0,0,0,0,85,63060,0,0,0,0,0,12,1,0,0,0,0,0,0,"Stormwind Valiant - On Script - Invoker Cast 'Quest Credit - Trigger - 01'"),
(@ENTRY,9,2,0,0,0,100,0,0,0,0,0,85,64836,0,0,0,0,0,12,1,0,0,0,0,0,0,"Stormwind Valiant - On Script - Invoker Cast '[DND] Bested Trigger Stormwind'"),
(@ENTRY,9,3,0,0,0,100,0,0,0,0,0,27,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Quest Credit ''"),
(@ENTRY,9,4,0,0,0,100,0,0,0,0,0,102,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Set Health Regeneration Off"),
(@ENTRY,9,5,0,0,0,100,0,2000,2000,0,0,66,0,0,0,0,0,0,7,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Set Orientation Invoker"),
(@ENTRY,9,6,0,0,0,100,0,0,0,0,0,1,2,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Say Line 2"),
(@ENTRY,9,7,0,0,0,100,0,5000,5000,0,0,24,0,0,0,0,0,0,1,0,0,0,0,0,0,0,"Stormwind Valiant - On Script - Evade");
```

If the script above is approved, it can be copied for the other valiants with modified:
 - entry ids
 - actionlist ids
 - spell [DND] Bested Trigger Stormwind based on race

Note to self:
Still need to check if valiants and champions have the same combat behaviour.

**Comment by Rushor** [comment]
I'm going to rewrite the script of those creatures to cpp to make shure, that they don't use autoattacks, but instead only http://www.wowhead.com/spell=62544/thrust .

**Comment by Rushor** [comment]
Scripts/Quest: The Grand Melee - Move Valiants & Champions script to Corescript

* the Valiants & Champions no properly reward the questitem after defeating
* the Valiants & Champions no longer meleeattack, but instead only use their ability SPELL_THRUST in meleerange
* update combatbehaviour of Valiants & Champions. Champions in general repeat all their spellcasts and events 3 seconds earlier then the Valiants + the Champions get 3 stacks of SPELL_DEFEND (62719), the Valiants only get 2 stacks of SPELL_DEFEND (62719).
* closes: https://github.com/dalaranwow/dalaran-wow/issues/4692
* commit: ab095bf2b4405d5e51086ea1ac20d437dabf5833

http://i.imgur.com/NzWtbjn.jpg

The issue was fixed by a developer and will be tested on the Public Test Realm in the future (PTR). The changes/fixes will be applied with the NEXT changelog on the liverealm in the future.
This means that the issue exists in game at the moment i wrote this message.
Only once the issue was marked with "Fixed - On Live Server", you can find the fix on you server.
Please watch http://www.dalaran-wow.com/news/ to see when the changelog and the fix was applied.
* commit:
